synedrion/src/uint/secret_signed.rs

```diff
@@ -125,34 +118,44 @@ where
         )
     }
 
+    /// Creates a [`SignedSecret`] from an unsigned value in range `[0, modulus)`,
+    /// treating the values greater than `modulus / 2` as negative ones (modulo `modulus`).
+    /// `modulus_bound` is the bit bound for the modulus; the bound of the result will be set to `modulus_bound - 1`
+    /// (since it is the bound for the absolute value).
+    ///
+    /// Returns `None` if the bound is too large, or if `abs(value)` is greater or equal to `2^bound`.
+    pub fn new_modulo(value: Secret<T>, modulus: &NonZero<T>, modulus_bound: u32) -> Option<Self> {
+        // We are taking a `bound` explicitly and not deriving it from the `modulus`
+        // because we want it to be the same across different runs, and the RSA modulus, being a product of two randoms,
+        // can have varying size (e.g. for two random 1024 bit primes it can be 2022-2024 bits long).
+
+        let positive = value;
+        let half_modulus = modulus.as_ref().wrapping_shr_vartime(1);
+        let is_negative = positive.expose_secret().ct_gt(&half_modulus);
+        // Can't define a `Sub<Secret>` for `Uint`, so have to re-wrap manually.
+        let negative = Secret::init_with(|| *modulus.as_ref() - positive.expose_secret()).wrapping_neg();
+        let value = Secret::<T>::conditional_select(&positive, &negative, is_negative);
+        Self::new_from_unsigned(value, modulus_bound - 1)
+    }
+
     /// Returns a truthy `Choice` if the absolute value is within the bit bound `bound`.
-    fn in_bound(&self, bound: u32) -> Choice {
+    fn is_in_bound(&self, bound: u32) -> Choice {
         let abs = self.abs();
         let mask = T::one().wrapping_neg().wrapping_shl_vartime(bound);
         let masked = abs & mask;
         masked.is_zero()
     }
 
-    /// Asserts that the absolute value is within the bit bound `bound`.
-    /// If that is the case, returns the value with the bound set to it.
-    pub fn ensure_bound(&self, bound: u32) -> CtOption<Self> {
-        let value = Self {
-            value: self.value.clone(),
-            bound,
-        };
-        CtOption::new(value, self.in_bound(bound))
-    }
-
-    /// Asserts that the value is within the interval the paper denotes as $\pm 2^exp$.
+    /// Asserts that the value is within the interval the paper denotes as $±2^exp$.
     /// Panics if it is not the case.
     ///
-    /// That is, the value must be within $[-2^{exp}, 2^{exp}]$
-    /// (See Section 2).
+    /// That is, the value must be within $[-2^{exp-1}+1, 2^{exp-1}]$
+    /// (See Section 3, Groups & Fields).
     pub fn assert_exponent_range(&self, exp: u32) {
-        let in_bound = self.in_bound(exp);
-        // Have to check for the ends of the range too
-        let is_end = self.abs().expose_secret().ct_eq(&(T::one() << exp));
-        assert!(bool::from(in_bound | is_end), "out of bounds $\\pm 2^{exp}$",)
+        let in_bound = self.is_in_bound(exp - 1);
```

Comment:
Maybe dumb question, but the new lower bound is `-2^{exp-1}+1` and I don't see the `+1` being checked for, can you spell it out for me?

Reply:
`is_in_bound()` checks that the value is under `2^bound` (that is, in range `[0, 2^bound-1]`). We're applying that to the absolute value, so we're checking that the original number is in `[-2^bound+1, 2^bound-1]`. `bound = exp - 1`.